File: cpp/vm/advm/parser.cpp
#include "parser.h"

Parser::Parser() {
    statementParseFns.insert(std::make_pair(TT_IF, &Parser::parseIfStatement));

    prefixParseFns.insert(std::make_pair(TT_IDENT, &Parser::parseIdent));
    prefixParseFns.insert(std::make_pair(TT_INT, &Parser::parseIntegerLiteral));
    prefixParseFns.insert(std::make_pair(TT_MINUS, &Parser::parsePrefixExpression));
    prefixParseFns.insert(std::make_pair(TT_BANG, &Parser::parsePrefixExpression));
    prefixParseFns.insert(std::make_pair(TT_TRUE, &Parser::parseBooleanExpression));
    prefixParseFns.insert(std::make_pair(TT_FALSE, &Parser::parseBooleanExpression));
    prefixParseFns.insert(std::make_pair(TT_LPAREN, &Parser::parseGroupedExpression));
    //prefixParseFns.insert(std::make_pair(TT_IF, [this]() { return parseIfStatement(); } ));
    prefixParseFns.insert(std::make_pair(TT_IF, &Parser::parseIfStatement));
    prefixParseFns.insert(std::make_pair(TT_NULL, &Parser::parseNullExpression));
    prefixParseFns.insert(std::make_pair(TT_STRING, &Parser::parseStringLiteral));
    prefixParseFns.insert(std::make_pair(TT_LBRACKET, &Parser::parseListLiteral));
    prefixParseFns.insert(std::make_pair(TT_LBRACE, &Parser::parseHashLiteral));

    prefixParseFns.insert(std::make_pair(TT_FUNC, &Parser::parseFunctionLiteral));

    infixParseFns.insert(std::make_pair(TT_PLUS, &Parser::parseInfixExpression));
    infixParseFns.insert(std::make_pair(TT_MINUS, &Parser::parseInfixExpression));
    infixParseFns.insert(std::make_pair(TT_MULTIPLY, &Parser::parseInfixExpression));
    infixParseFns.insert(std::make_pair(TT_DIVIDE, &Parser::parseInfixExpression));
    infixParseFns.insert(std::make_pair(TT_GT, &Parser::parseInfixExpression));
    infixParseFns.insert(std::make_pair(TT_LT, &Parser::parseInfixExpression));
    infixParseFns.insert(std::make_pair(TT_GTE, &Parser::parseInfixExpression));
    infixParseFns.insert(std::make_pair(TT_LTE, &Parser::parseInfixExpression));
    infixParseFns.insert(std::make_pair(TT_EQUALS, &Parser::parseInfixExpression));
    infixParseFns.insert(std::make_pair(TT_LBRACKET, &Parser::parseIndexExpression));
    infixParseFns.insert(std::make_pair(TT_LPAREN, &Parser::parseCallExpression));
}

PrecedenceType Parser::currentPrecedence() {
    return precedenceMap[currentToken.type];
}

PrecedenceType Parser::peekPrecedence() {
    return precedenceMap[peekToken.type];
}

bool Parser::currentTokenIs(TokenType type) {
    return currentToken.type == type;
}

bool Parser::peekTokenIs(TokenType type) {
    return peekToken.type == type;
}

bool Parser::expectPeek(TokenType type) {
    if (peekTokenIs(type)) {
        nextToken();
        return true;
    }
    std::cout << "expect token error: " << tokenTypeConverter[type] << " not found\n";
    return false;
}

void Parser::load(std::string source) {
    lexer.load(source);
    nextToken();
    nextToken();
}

void Parser::buildProgramStatement(ASTProgram *program) {
    while(currentToken.type != TT_EOF) {
        // std::cout << currentToken.toString() << "\n";
        ASTNode* stmt = parseStatement();
        program->statements.push_back(stmt);
        nextToken();
    }
}

void Parser::nextToken() {
    currentToken = peekToken;
    peekToken = lexer.nextToken();
}

ASTNode* Parser::parseStatement() {
    if (currentToken.type == TT_LET)
        return parseLetStatement();
    if (currentToken.type == TT_RETURN)
        return parseReturnStatement();
    /*if (currentToken.type == TT_BREAK)
        return parseBreakStatement();
    if (currentToken.type == TT_CONTINUE)
        return parseContinueStatement();
    if (currentToken.type == TT_MULTICOMMENT)
        return parseComment();
    if (currentToken.type == TT_SINGLECOMMENT)
        return parseSingleLineComment();
    if (currentToken.type == TT_IF)
        return parseIfStatement();*/
    return parseExpressionStatement();
}

ASTNode* Parser::parseIdent() {
    ASTIdentifier *ident = new ASTIdentifier(currentToken, currentToken.stringLiteral);
    return ident;
}

ASTNode* Parser::parseLetStatement() {
    ASTLetStatement* stmt = new ASTLetStatement(currentToken);

    if (!expectPeek(TT_IDENT)) {
        //delete stmt;
        free_memory_ASTNode(stmt);
        return NULL;
    }

    stmt->name = ASTIdentifier(currentToken, currentToken.stringLiteral);

    if (!expectPeek(TT_ASSIGN)) {
        //delete stmt;
        free_memory_ASTNode(stmt);
        return NULL;
    }

    nextToken();
    stmt->value = parseExpression(PT_LOWEST);
    if (currentTokenIs(TT_SEMICOLON)) {
        nextToken();
    }
    return stmt;
}

ASTNode* Parser::parseReturnStatement() {
    ASTReturnStatement* stmt = new ASTReturnStatement(currentToken);
    nextToken();
    stmt->value = parseExpression(PT_LOWEST);
    if (peekTokenIs(TT_RBRACE) || peekTokenIs(TT_SEMICOLON) || peekTokenIs(TT_EOF)) {
        return stmt;
    }
    return stmt;
}

ASTNode* Parser::parseBreakStatement() {
    return NULL;
}

ASTNode* Parser::parseContinueStatement() {
    return NULL;
}

ASTNode* Parser::parseComment() {
    return NULL;
}

ASTNode* Parser::parseSingleLineComment() {
    return NULL;
}

ASTNode* Parser::parseIntegerLiteral() {
    ASTInteger* integer_node = new ASTInteger();
    int val = stoi(currentToken.stringLiteral);
    integer_node->token = currentToken;
    integer_node->value = val;
    return integer_node;
}

ASTNode* Parser::parseBooleanExpression() {
    if (currentToken.type == TT_TRUE) {
        ASTBoolean* boolean_node = new ASTBoolean();
        boolean_node->token = currentToken;
        boolean_node->value = true;
        return boolean_node;
    }
    if (currentToken.type == TT_FALSE) {
        ASTBoolean* boolean_node = new ASTBoolean();
        boolean_node->token = currentToken;
        boolean_node->value = false;
        return boolean_node;
    }
    return NULL;
}

ASTNode* Parser::parseGroupedExpression() {
    nextToken();
    ASTNode *expr = parseExpression(PT_LOWEST);
    if (!expectPeek(TT_RPAREN)) {
        return NULL;
    }
    return expr;
}

ASTNode* Parser::parseIfStatement() {
    ASTIfStatement* expr = new ASTIfStatement(currentToken);
    if (!expectPeek(TT_LPAREN)) {
        free_memory_ASTNode(expr);
        return NULL;
    }
    nextToken();
    expr->condition = parseExpression(PT_LOWEST);
    if (!expectPeek(TT_RPAREN)) {
        free_memory_ASTNode(expr);
        return NULL;
    }
    if (!expectPeek(TT_LBRACE)) {
        expr->consequence = parseSingleBlockStatement();
    } else {
        expr->consequence = parseBlockStatement();
    }

    if (peekTokenIs(TT_ELSE)) {
        nextToken();
        if (!expectPeek(TT_LBRACE)) {
            expr->alternative = parseSingleBlockStatement();
        } else {
            expr->alternative = parseBlockStatement();
        }
    }
    return expr;
}

ASTNode* Parser::parseSingleBlockStatement() {
    ASTBlockStatement* block = new ASTBlockStatement(currentToken);
    nextToken();
    ASTNode* stmt = (ASTNode*)parseStatement();
    if (stmt) {
        block->statements.push_back(stmt);
    }
    return block;

}

ASTNode* Parser::parseBlockStatement() {
    ASTBlockStatement* block = new ASTBlockStatement(currentToken);
    nextToken();
    while(!currentTokenIs(TT_RBRACE) && !currentTokenIs(TT_EOF)) {
        ASTNode* stmt = (ASTNode*)parseStatement();
        if (stmt) {
            block->statements.push_back(stmt);
        }
        nextToken();
    }
    return block;
}

ASTNode* Parser::parseNullExpression() {
    ASTNullExpression *expr = new ASTNullExpression();
    nextToken();
    return expr;
}

ASTNode* Parser::parseStringLiteral() {
    ASTString *expr = new ASTString(currentToken, currentToken.stringLiteral);
    return expr;
}

ASTNode* Parser::parseListLiteral() {
    ASTList* expr = new ASTList(currentToken);
    expr->elements = parseListExpressions();
    return expr;
}

std::vector<ASTNode*> Parser::parseListExpressions() {
    std::vector<ASTNode*> elements;
    if (peekTokenIs(TT_RBRACKET)) {
        nextToken();
        return elements;
    }
    nextToken();
    ASTNode* element = parseExpression(PT_LOWEST);
    //Ad_INCREF(element); // this might be needed
    elements.push_back(element);
    while(peekTokenIs(TT_COMMA)) {
        nextToken();
        nextToken();
        element = parseExpression(PT_LOWEST);
        //Ad_INCREF(element); // this might be needed
        elements.push_back(element);
    }
    if (!expectPeek(TT_RBRACKET)) {
        std::vector<ASTNode*> empty;
        return empty;
    }
    return elements;
}

ASTNode* Parser::parseHashLiteral() {
    ASTHash* hash = new ASTHash(currentToken);
    while (!peekTokenIs(TT_RBRACE)) {
        nextToken();
        ASTNode* key = parseExpression(PT_LOWEST);
        if (!expectPeek(TT_COLON)) {
            return NULL;
        }
        nextToken();
        ASTNode* value = parseExpression(PT_LOWEST);
        //Ad_INCREF(key);
        //Ad_INCREF(value);
        hash->pairs.insert(std::make_pair(key, value));
        if (!peekTokenIs(TT_RBRACE) && !expectPeek(TT_COMMA)) {
            return NULL;
        }
    }
    if (!expectPeek(TT_RBRACE)) {
        return NULL;
    }
    return hash;
}

ASTNode* Parser::parseFunctionLiteral() {
    ASTFunctionLiteral* fun_lit = new ASTFunctionLiteral(currentToken);
    if (!expectPeek(TT_LPAREN)) {
        //delete fun_lit;
        free_memory_ASTNode(fun_lit);
        return NULL;
    }
    std::pair<std::vector<ASTNode*>, std::vector<ASTNode*>> res = parseFunctionParameters();
    fun_lit->parameters = res.first;
    fun_lit->default_params = res.second;
    if (!expectPeek(TT_LBRACE)) {
        //delete fun_lit;
        free_memory_ASTNode(fun_lit);
        return NULL;
    }
    fun_lit->body = parseBlockStatement();
    Ad_INCREF(fun_lit->body);
    return fun_lit;
}

std::pair<std::vector<ASTNode*>, std::vector<ASTNode*>> Parser::parseFunctionParameters() {
    std::vector<ASTNode*> identifiers;
    std::vector<ASTNode*> defaultParams;
    if (peekTokenIs(TT_RPAREN)) {
        nextToken();
        return std::make_pair(identifiers, defaultParams);
    }
    nextToken();
    ASTIdentifier* ident = new ASTIdentifier(currentToken, currentToken.stringLiteral);
    identifiers.push_back(ident);
    Ad_INCREF(ident);
    if (peekTokenIs(TT_ASSIGN)) {
        nextToken();
        nextToken();
        ASTNode *value = parseExpression(PT_LOWEST);
        defaultParams.push_back(value);
        Ad_INCREF(value);
    }
    while (peekTokenIs(TT_COMMA)) {
        nextToken();
        nextToken();
        ident = new ASTIdentifier(currentToken, currentToken.stringLiteral);
        identifiers.push_back(ident);
        Ad_INCREF(ident);
        if (peekTokenIs(TT_ASSIGN)) {
            nextToken();
            nextToken();
            ASTNode *value = parseExpression(PT_LOWEST);
            defaultParams.push_back(value);
            Ad_INCREF(value);
        } else {
            if (defaultParams.size() > 0) {
                std::cout << "ERROR: no positional arguments allowed after defaulted params!";
            }
        }
    }
    if (!expectPeek(TT_RPAREN)) {
        std::vector<ASTNode*> empty; // i don't like this, it should be NULL
        return std::make_pair(empty, empty);
    }
    return std::make_pair(identifiers, defaultParams);
}

ASTNode* Parser::parseCallExpression(ASTNode* left) {
    ASTCallExpression* expr = new ASTCallExpression(currentToken, left);
    std::pair<std::vector<ASTNode*>, std::vector<ASTNode*>> res = parseCallArguments();
    expr->arguments = res.first;
    expr->kw_args = res.second;
    return expr;
}

std::pair<std::vector<ASTNode*>, std::vector<ASTNode*>> Parser::parseCallArguments() {
    std::vector<ASTNode*> args;
    std::vector<ASTNode*> kw_args;
    if (peekTokenIs(TT_RPAREN)) {
        nextToken();
        return std::make_pair(args, kw_args);
    }
    nextToken();
    ASTNode *expr1 = parseExpression(PT_LOWEST);
    if (expr1->type == AT_ASSIGN_STATEMENT) {
        kw_args.push_back(expr1);
    } else {
        args.push_back(expr1);
    }
    while (peekTokenIs(TT_COMMA)) {
        nextToken();
        nextToken();
        ASTNode *expr1 = parseExpression(PT_LOWEST);
        if (expr1->type == AT_ASSIGN_STATEMENT) {
            kw_args.push_back(expr1);
        } else {
            args.push_back(expr1);
        }
    }
    if (!expectPeek(TT_RPAREN)) {
        std::vector<ASTNode*> empty; // i don't like this, it should be NULL
        return std::make_pair(empty, empty);
    }
    return std::make_pair(args, kw_args);
}

ASTNode* Parser::parseIndexExpression(ASTNode* left) {
    ASTIndexExpression* expr = new ASTIndexExpression(currentToken, left);
    nextToken();

    if (currentTokenIs(TT_COLON)) {
        ASTNode *index = new ASTNullExpression();
        expr->index = index;
    } else {
        ASTNode *index = parseExpression(PT_LOWEST);
        expr->index = index;
        nextToken();
        if (currentTokenIs(TT_RBRACKET)) {
            return expr;
        }
    }

    nextToken();

    if (currentTokenIs(TT_COLON)) {
        ASTNode *indexEnd = new ASTNullExpression();
        expr->indexEnd = indexEnd;
    } else {
        ASTNode *indexEnd = parseExpression(PT_LOWEST);
        expr->indexEnd = indexEnd;
        nextToken();
        if (currentTokenIs(TT_RBRACKET)) {
            ASTNode *step = new ASTInteger(currentToken, 1);
            expr->step = step;
            return expr;
        }
    }

    nextToken();

    if (currentTokenIs(TT_RBRACKET)) {
        ASTNode *step = new ASTInteger(currentToken, 1);
        expr->step = step;
        return expr;
    } else {
        ASTNode *step = parseExpression(PT_LOWEST);
        nextToken();
        expr->step = step;
    }

    if (!currentTokenIs(TT_RBRACKET)) {
        return NULL;
    }

    return expr;
}

ASTNode* Parser::parsePrefixExpression() {
    ASTPrefixExpression *expr = new ASTPrefixExpression(currentToken, currentToken.stringLiteral);
    nextToken();
    expr->right = parseExpression(PT_PREFIX);
    return expr;
}

ASTNode* Parser::parseInfixExpression(ASTNode* left) {
    ASTInfixExpression* expr = new ASTInfixExpression();
    expr->token = currentToken;
    expr->_operator = currentToken.stringLiteral;
    expr->left = left;
    PrecedenceType preced = currentPrecedence();
    nextToken();
    expr->right = parseExpression(preced);
    return expr;
}

ASTNode* Parser::parseExpressionStatement() {
    ASTExpressionStatement* stmt = new ASTExpressionStatement(currentToken);
    stmt->expression = parseExpression(PT_LOWEST);
    if (peekTokenIs(TT_SEMICOLON)) {
        nextToken();
    }
    return stmt;
}

ASTNode* Parser::parseExpression(PrecedenceType precedence) {
    if (prefixParseFns.find(currentToken.type) == prefixParseFns.end()) {
        return NULL;
    }
    PrefixCallback prefix = prefixParseFns[currentToken.type];

    ASTNode* leftExp = (this->*prefix)();

    while(!peekTokenIs(TT_SEMICOLON) && (precedence < peekPrecedence())) {
        if (infixParseFns.find(peekToken.type) == infixParseFns.end()) {
            return leftExp;
        }
        InfixCallback infix = infixParseFns[peekToken.type];
        nextToken();
        leftExp = (this->*infix)(leftExp);
    }
    return leftExp;
}

File: python/VMPrototype1/parser.py
from typing import Optional

from ad_ast import ASTProgram, ASTNode, ASTExpressionStatement, ASTInteger, ASTInfixExpression, ASTPrefixExpression, \
    ASTBoolean, ASTIfExpression, ASTBlockStatement, ASTNullExpression, ASTLetStatement, ASTIdentifier, ASTString, \
    ASTList, ASTIndexExpression, ASTHash, ASTFunctionLiteral, ASTCallExpression, StatementType, ASTReturnStatement, \
    ASTClassStatement, ASTDefStatement, ASTAssignStatement, ASTWhileExpression, ASTMemberAccess, ASTThisExpression
from lexer import Lexer
from precedence_type import PrecedenceType, precedences
from token_type import TokenType


class Parser:
    def __init__(self):
        self.source = None
        self.lexer = Lexer()
        self.current_token = None
        self.peek_token = None
        self.statement_parse_fns = {
            TokenType.IF: self.parse_if_statement,
            TokenType.DEF: self.parse_def_statement
        }

        self.prefix_parse_fns = {
            TokenType.IDENT: self.parse_identifier,
            TokenType.INT: self.parse_integer_literal,
            TokenType.TRUE: self.parse_boolean,
            TokenType.FALSE: self.parse_boolean,
            TokenType.BANG: self.parse_prefix_expression,
            TokenType.MINUS: self.parse_prefix_expression,
            TokenType.LPAREN: self.parse_grouped_expression,
            TokenType.IF: self.parse_if_statement,
            TokenType.NULL: self.parse_null_expression,
            TokenType.STRING: self.parse_string_literal,
            TokenType.LBRACKET: self.parse_list_literal,
            TokenType.LBRACE: self.parse_hash_literal,
            TokenType.FUNC: self.parse_func_literal,
            TokenType.DEF: self.parse_def_statement,
            TokenType.CLASS: self.parse_class_statement,
            TokenType.WHILE: self.parse_while_expression,
            TokenType.THIS: self.parse_this_expression
        }

        self.infix_parse_fns = {
            TokenType.PLUS: self.parse_infix_expression,
            TokenType.MINUS: self.parse_infix_expression,
            TokenType.ASTERISK: self.parse_infix_expression,
            TokenType.SLASH: self.parse_infix_expression,
            TokenType.EQ: self.parse_infix_expression,
            TokenType.NOT_EQ: self.parse_infix_expression,
            TokenType.GT: self.parse_infix_expression,
            TokenType.LT: self.parse_infix_expression,
            TokenType.GTE: self.parse_infix_expression,
            TokenType.LTE: self.parse_infix_expression,
            TokenType.LBRACKET: self.parse_index_expression,
            TokenType.LPAREN: self.parse_call_expression,
            TokenType.ASSIGN: self.parse_assign_expression,
            TokenType.DOT: self.parse_member_access
        }

    def load(self, source):
        self.source = source
        self.lexer.load(source)
        self.next_token()
        self.next_token()

    def build_program_statements(self, program: ASTProgram):
        while self.current_token.token_type != TokenType.EOF:
            # print(self.current_token)
            stmt = self.parse_statement()
            if stmt:
                program.statements.append(stmt)
            self.next_token()

    def next_token(self):
        self.current_token = self.peek_token
        self.peek_token = self.lexer.next_token()

    def current_token_is(self, token_type : TokenType):
        return self.current_token.token_type == token_type

    def peek_token_is(self, token_type: TokenType):
        return self.peek_token.token_type == token_type

    def expect_peek(self, token_type: TokenType) -> bool:
        if self.peek_token_is(token_type):
            self.next_token()
            return True
        else:
            #self.errors.append('ERROR: i was expecting a different token here: ' + token_type)
            print('ERROR: i was expecting a different token here: ' + str(token_type))
            return False

    def peek_precedence(self):
        if self.peek_token.token_type in precedences:
            return precedences[self.peek_token.token_type]
        return PrecedenceType.LOWEST

    def current_precedence(self):
        if self.current_token.token_type in precedences:
            return precedences[self.current_token.token_type]
        return PrecedenceType.LOWEST

    def parse_statement(self) -> Optional[ASTNode]:
        #if self.current_token.token_type in self.statement_parse_fns:
        #    return self.statement_parse_fns[self.current_token.token_type]()
        if self.current_token.token_type == TokenType.LET:
            return self.parse_let_expression()
        if self.current_token.token_type == TokenType.RETURN:
            return self.parse_return_statement()
        return self.parse_expression_statement()

    def parse_identifier(self) -> ASTNode:
        stmt = ASTIdentifier(token=self.current_token, value=self.current_token.literal)
        return stmt

    def parse_integer_literal(self) -> ASTNode:
        stmt = ASTInteger(token=self.current_token, value=int(self.current_token.literal))
        return stmt

    def parse_boolean(self) -> ASTNode:
        stmt = ASTBoolean(token=self.current_token, value=bool(self.current_token.token_type == TokenType.TRUE))
        return stmt

    def parse_expression_statement(self) -> Optional[ASTNode]:
        stmt = ASTExpressionStatement(token=self.current_token)
        stmt.expression = self.parse_expression(PrecedenceType.LOWEST)
        if self.peek_token_is(TokenType.SEMICOLON):
            self.next_token()
        return stmt

    def parse_prefix_expression(self) -> ASTNode:
        expr = ASTPrefixExpression(token=self.current_token, operator=self.current_token.literal)
        self.next_token()
        expr.right = self.parse_expression(PrecedenceType.PREFIX)
        return expr

    def parse_grouped_expression(self) -> Optional[ASTNode]:
        self.next_token()
        expr = self.parse_expression(PrecedenceType.LOWEST)
        if not self.expect_peek(TokenType.RPAREN):
            return None
        return expr

    def parse_if_statement(self) -> Optional[ASTNode]:
        expr = ASTIfExpression(token=self.current_token)
        if not self.expect_peek(TokenType.LPAREN):
            return None
        self.next_token()
        expr.condition = self.parse_expression(PrecedenceType.LOWEST)
        if not self.expect_peek(TokenType.RPAREN):
            return None
        if not self.expect_peek(TokenType.LBRACE):
            expr.consequence = self.parse_single_block_statement()
        else:
            expr.consequence = self.parse_block_statement()
        if self.peek_token_is(TokenType.ELSE):
            self.next_token()
            if not self.expect_peek(TokenType.LBRACE):
                expr.alternative = self.parse_single_block_statement()
            else:
                expr.alternative = self.parse_block_statement()
        return expr

    def parse_block_statement(self):
        block = ASTBlockStatement(token=self.current_token)
        self.next_token()
        while (not self.current_token_is(TokenType.RBRACE)) and (not self.current_token_is(TokenType.EOF)):
            stmt = self.parse_statement()
            if stmt:
                block.statements.append(stmt)
            self.next_token()
        return block

    def parse_single_block_statement(self):
        block = ASTBlockStatement(token=self.current_token)
        self.next_token()
        stmt = self.parse_statement()
        block.statements.append(stmt)
        return block

    def parse_while_expression(self):
        expr = ASTWhileExpression(token=self.current_token)
        if not self.expect_peek(TokenType.LPAREN):
            return None
        self.next_token()
        expr.condition = self.parse_expression(PrecedenceType.LOWEST)
        if not self.expect_peek(TokenType.RPAREN):
            return None
        if not self.expect_peek(TokenType.LBRACE):
            expr.block = self.parse_single_block_statement()
        else:
            expr.block = self.parse_block_statement()
        return expr

    def parse_null_expression(self) -> ASTNode:
        expr = ASTNullExpression(token=self.current_token)
        self.next_token()
        return expr

    def parse_this_expression(self):
        expr = ASTThisExpression(token=self.current_token)
        return expr

    def parse_string_literal(self) -> ASTNode:
        return ASTString(token=self.current_token, value=str(self.current_token.literal))

    def parse_list_literal(self) -> ASTNode:
        expr = ASTList(self.current_token)
        expr.elements = self.parse_expression_list()
        return expr

    def parse_hash_literal(self) -> Optional[ASTNode]:
        node = ASTHash(token=self.current_token)
        node.pairs = {}
        while not self.peek_token_is(TokenType.RBRACE):
            self.next_token()
            key = self.parse_expression(PrecedenceType.LOWEST)
            if not self.expect_peek(TokenType.COLON):
                return None
            self.next_token()
            value = self.parse_expression(PrecedenceType.LOWEST)
            node.pairs[key] = value
            if not self.peek_token_is(TokenType.RBRACE) and not self.expect_peek(TokenType.COMMA):
                return None
        if not self.expect_peek(TokenType.RBRACE):
            return None
        return node

    def parse_expression_list(self):
        elements = []
        if self.peek_token_is(TokenType.RBRACKET):
            self.next_token()
            return elements
        self.next_token()
        elements.append(self.parse_expression(PrecedenceType.LOWEST))
        while self.peek_token_is(TokenType.COMMA):
            self.next_token()
            self.next_token()
            elements.append(self.parse_expression(PrecedenceType.LOWEST))
        if not self.expect_peek(TokenType.RBRACKET):
            return None
        return elements

    def parse_call_expression(self, left: ASTNode) -> Optional[ASTNode]:
        expr = ASTCallExpression(self.current_token, left)
        res = self.parse_call_arguments()
        expr.arguments = res[0]
        expr.kw_args = res[1]
        return expr

    def parse_call_arguments(self):
        args = []
        kw_args = []
        if self.peek_token_is(TokenType.RPAREN):
            self.next_token()
            return args, kw_args
        self.next_token()
        expr1 = self.parse_expression(PrecedenceType.LOWEST)
        if expr1.statement_type == StatementType.ASSIGN_STATEMENT:
            kw_args.append(expr1)
        else:
            args.append(expr1)
        while self.peek_token_is(TokenType.COMMA):
            self.next_token()
            self.next_token()
            if self.peek_token_is(TokenType.ASSIGN):
                kw_args.append(self.parse_expression(PrecedenceType.LOWEST))
            else:
                args.append(self.parse_expression(PrecedenceType.LOWEST))
        if not self.expect_peek(TokenType.RPAREN):
            return [], {}
        return args, kw_args

    def parse_index_expression(self, left: ASTNode) -> Optional[ASTNode]:
        expr = ASTIndexExpression(token=self.current_token, left=left)
        self.next_token()
        expr.index = self.parse_expression(PrecedenceType.LOWEST)
        if not self.expect_peek(TokenType.RBRACKET):
            return None
        return expr

    def parse_assign_expression(self, left: ASTNode) -> ASTNode:
        stmt = ASTAssignStatement(token=self.current_token)
        stmt.name = left
        self.next_token()
        stmt.value = self.parse_expression(PrecedenceType.LOWEST)
        if self.current_token_is(TokenType.SEMICOLON):
            self.next_token()
        return stmt

    def parse_member_access(self, left: ASTNode) -> ASTNode:
        member_access = ASTMemberAccess(token=self.current_token)
        self.next_token()
        right = ASTIdentifier(token=self.current_token, value=self.current_token.literal)
        member_access.owner = left
        member_access.member = right

        if self.peek_token_is(TokenType.LPAREN):
            self.next_token()
            res = self.parse_call_arguments()
            member_access.arguments = res[0]
            member_access.kw_args = res[1]
            member_access.is_method = True
        else:
            member_access.arguments = []
            member_access.kw_args = []
            member_access.is_method = False
        return member_access

    def parse_let_expression(self):
        stmt = ASTLetStatement(token=self.current_token)
        if not self.expect_peek(TokenType.IDENT):
            return None
        stmt.name = ASTIdentifier(token=self.current_token, value=self.current_token.literal)
        if not self.expect_peek(TokenType.ASSIGN):
            return None
        self.next_token()
        stmt.value = self.parse_expression(PrecedenceType.LOWEST)
        if stmt.value.statement_type == StatementType.FUNCTION_LITERAL:
            stmt.value.name = stmt.name.value
        if self.current_token_is(TokenType.SEMICOLON):
            self.next_token()
        return stmt

    def parse_func_literal(self):
        func = ASTFunctionLiteral(token=self.current_token)
        if not self.expect_peek(TokenType.LPAREN):
            return None
        res = self.parse_function_parameters()
        func.parameters, func.default_params = res[0], res[1]
        if not self.expect_peek(TokenType.LBRACE):
            return None
        func.body = self.parse_block_statement()
        return func

    def parse_def_statement(self):
        stmt = ASTDefStatement(token=self.current_token)
        self.next_token()
        name = ASTIdentifier(token=self.current_token, value=self.current_token.literal)
        stmt.name = name
        if not self.expect_peek(TokenType.LPAREN):
            # this should return an error object
            return None
        res = self.parse_function_parameters()
        stmt.parameters, stmt.default_params = res[0], res[1]
        if not self.expect_peek(TokenType.LBRACE):
            return None
        stmt.body = self.parse_block_statement()
        return stmt

    def parse_class_statement(self):
        expr = ASTClassStatement(token=self.current_token)
        self.next_token()
        name = self.parse_identifier()
        expr.name = name
        self.next_token()
        expr.attributes = []
        expr.methods = []
        expr.inherit_from = []

        if self.current_token_is(TokenType.COLON):
            self.next_token()
            while not self.current_token_is(TokenType.LBRACE):
                parent = self.parse_identifier()
                expr.inherit_from.append(parent)
                self.next_token()

        while not self.current_token_is(TokenType.RBRACE):
            if self.current_token_is(TokenType.DEF):
                stmt = self.parse_def_statement()
                expr.methods.append(stmt)
            if self.current_token_is(TokenType.FUN):
                stmt = self.parse_def_statement()
                expr.methods.append(stmt)
            if self.current_token_is(TokenType.METHOD):
                stmt = self.parse_def_statement()
                expr.methods.append(stmt)
            elif self.current_token_is(TokenType.IDENT):
                stmt = self.parse_expression_statement()
                expr.attributes.append(stmt)
            self.next_token()
        return expr

    def parse_function_parameters(self):
        identifiers = []
        default_params = []
        if self.peek_token_is(TokenType.RPAREN):
            self.next_token()
            return identifiers, default_params
        self.next_token()
        ident = ASTIdentifier(token=self.current_token, value=self.current_token.literal)
        identifiers.append(ident)
        if self.peek_token_is(TokenType.ASSIGN):
            self.next_token()
            self.next_token()
            value = self.parse_expression(PrecedenceType.LOWEST)
            default_params.append(value)
        while self.peek_token_is(TokenType.COMMA):
            self.next_token()
            self.next_token()
            ident = ASTIdentifier(token=self.current_token, value=self.current_token.literal)
            identifiers.append(ident)
            if self.peek_token_is(TokenType.ASSIGN):
                self.next_token()
                self.next_token()
                value = self.parse_expression(PrecedenceType.LOWEST)
                default_params.append(value)
            else:
                if len(default_params) > 0 and default_params[-1] is not None:
                    raise Exception("problem!!!")
                #default_params.append(None)
        if not self.expect_peek(TokenType.RPAREN):
            return [], []
        return identifiers, default_params

    def parse_return_statement(self):
        stmt = ASTReturnStatement(token=self.current_token)
        self.next_token()
        stmt.value = self.parse_expression(PrecedenceType.LOWEST)
        if self.peek_token_is(TokenType.SEMICOLON) or self.peek_token_is(TokenType.RBRACE) or self.peek_token_is(TokenType.EOF):
            return stmt
        while not self.current_token_is(TokenType.SEMICOLON) and not self.current_token_is(TokenType.RBRACE) and not self.current_token_is(TokenType.EOF):
            self.next_token()
        return stmt

    def parse_infix_expression(self, left: ASTNode) -> ASTNode:
        expr = ASTInfixExpression(token=self.current_token, operator=self.current_token.literal, left=left)
        precedence = self.current_precedence()
        self.next_token()
        expr.right = self.parse_expression(precedence)
        return expr

    def parse_expression(self, precedence) -> Optional[ASTNode]:
        if self.current_token.token_type not in self.prefix_parse_fns:
            return None
        prefix = self.prefix_parse_fns[self.current_token.token_type]
        left_exp: Optional[ASTNode] = prefix()
        while (not self.peek_token_is(TokenType.SEMICOLON)) and precedence < self.peek_precedence():
            if self.peek_token.token_type not in self.infix_parse_fns:
                return left_exp
            infix = self.infix_parse_fns[self.peek_token.token_type]
            self.next_token()
            left_exp = infix(left_exp)
        return left_exp
